File: matrix_algorithms/src/matrices/Algoritmo15_V4_ParallelBlock.java
package matrices;

import java.math.BigInteger;
import java.util.concurrent.*;

/**
 * Algoritmo 15: V.4 Parallel Block
 * Tipo: Iterativo por bloques con transpuesta y paralelismo
 * Descripcion: Version paralela de V.3. Combina las tres tecnicas:
 *   1. Transpuesta de B para acceso cache-friendly
 *   2. Multiplicacion por bloques para reuso de cache
 *   3. Paralelismo con particion de filas sin sincronizacion
 * Es el algoritmo mas completo de la familia de bloques.
 * Complejidad: O(n^3 / p) con mejor localidad de cache
 */
public class Algoritmo15_V4_ParallelBlock {

    static final int BLOQUE = 64;
    static final int HILOS  = Runtime.getRuntime().availableProcessors();

    public static BigInteger[][] multiplicar(BigInteger[][] A, BigInteger[][] B) throws Exception {
        int n = A.length;

        // Transponer B
        BigInteger[][] BT = new BigInteger[n][n];
        for (int i=0;i<n;i++) for (int j=0;j<n;j++) BT[j][i] = B[i][j];

        BigInteger[][] C = new BigInteger[n][n];
        for (int i=0;i<n;i++) for (int j=0;j<n;j++) C[i][j]=BigInteger.ZERO;

        ExecutorService pool = Executors.newFixedThreadPool(HILOS);
        java.util.List<Future<?>> tareas = new java.util.ArrayList<>();

        // Particion de filas sin solapamiento
        int filasPorHilo = Math.max(1, n / HILOS);
        for (int h=0; h<HILOS; h++) {
            final int filaInicio = h * filasPorHilo;
            final int filaFin    = (h == HILOS-1) ? n : filaInicio + filasPorHilo;
            if (filaInicio >= n) break;
            tareas.add(pool.submit(() -> {
                for (int ii=filaInicio; ii<filaFin; ii+=BLOQUE)
                    for (int jj=0; jj<n; jj+=BLOQUE)
                        for (int kk=0; kk<n; kk+=BLOQUE) {
                            int iM=Math.min(ii+BLOQUE,filaFin);
                            int jM=Math.min(jj+BLOQUE,n), kM=Math.min(kk+BLOQUE,n);
                            for (int i=ii; i<iM; i++)
                                for (int j=jj; j<jM; j++) {
                                    BigInteger s = BigInteger.ZERO;
                                    for (int k=kk; k<kM; k++)
                                        s = s.add(A[i][k].multiply(BT[j][k]));
                                    C[i][j] = C[i][j].add(s);
                                }
                        }
            }));
        }
        pool.shutdown();
        for (Future<?> f : tareas) f.get();
        return C;
    }
}

File: matrix_algorithms_python/algoritmo15_V4_parallel_block.py
# Algoritmo 15: V.4 Parallel Block con transpuesta
# Complejidad: O(n^3 / p) con mejor localidad de cache
# Combina las tres tecnicas:
#   1. Transpuesta de B para acceso cache-friendly
#   2. Multiplicacion por bloques para reuso de cache
#   3. Paralelismo con particion de filas sin sincronizacion

import threading
import os

BLOQUE = 64
HILOS  = os.cpu_count()

def multiplicar(A, B):
    n = len(A)

    # transponer B
    BT = [[B[i][j] for i in range(n)] for j in range(n)]

    C = [[0] * n for _ in range(n)]

    def procesar_rango(fila_inicio, fila_fin):
        for ii in range(fila_inicio, fila_fin, BLOQUE):
            for jj in range(0, n, BLOQUE):
                for kk in range(0, n, BLOQUE):
                    iM = min(ii + BLOQUE, fila_fin)
                    jM = min(jj + BLOQUE, n)
                    kM = min(kk + BLOQUE, n)
                    for i in range(ii, iM):
                        for j in range(jj, jM):
                            s = 0
                            for k in range(kk, kM):
                                s += A[i][k] * BT[j][k]  # ambos accesos por fila
                            C[i][j] += s

    # particion de filas sin solapamiento
    filas_por_hilo = max(1, n // HILOS)
    hilos = []
    for h in range(HILOS):
        inicio = h * filas_por_hilo
        fin    = n if h == HILOS - 1 else inicio + filas_por_hilo
        if inicio >= n:
            break
        t = threading.Thread(target=procesar_rango, args=(inicio, fin))
        hilos.append(t)
        t.start()
    for t in hilos:
        t.join()
    return C
